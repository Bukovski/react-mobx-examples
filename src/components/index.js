export { TodoNew } from './todo-new.component';
export { TodoList } from './todo-list.component';
export { TodoItem } from './todo-item.component';
