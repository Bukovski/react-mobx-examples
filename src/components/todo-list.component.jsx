import React from 'react';
import { TodoItem } from "./index";
import { useStore } from "../helpers/store-context.helper";


export const TodoList = () => {
	const todoList = useStore();
	
	return (
		<div className="todo-list">
			<ul className="open-todos todos align">
				<span>Open Todos</span>
				{
					todoList.openTodos.map(todo =>
						<TodoItem key={ todo.id } todo={ todo } />
					)
				}
			</ul>
			<ul className="finished-todos todos align">
				<span>Finished Todos</span>
				{
					todoList.finishedTodos.map(todo =>
						<TodoItem key={ todo.id } todo={ todo } />
					)
				}
			</ul>
		</div>
	)
};