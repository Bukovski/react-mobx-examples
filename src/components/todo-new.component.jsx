import React, { useState } from 'react';


export const TodoNew = () => {
	const [ todo, setTodo ] = useState('');
	
	const addTodo = () => {
		setTodo('');
	};
	
	return (
		<div className="add">
			<input
				type="text"
				name="add"
				placeholder="Add item..."
				value={ todo }
				onChange={ (e) => setTodo(e.target.value) }
			/>
				<div className="input-buttons" onClick={ addTodo }>
					<button className="add-todo">
						<i className="fas fa-plus add plus-icon"/>
					</button>
				</div>
		</div>
	)
};
