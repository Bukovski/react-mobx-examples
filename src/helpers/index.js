export { StoreContext, StoreProvider, useStore } from './store-context.helper';
export { randomId } from './random-id.helper';
export { onEnterPress } from './use-enter.helper';
