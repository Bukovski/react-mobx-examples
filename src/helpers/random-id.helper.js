export const randomId = () => {
	return '_'
		+ (Date.now().toString(36)
			+ Math.random().toString(36).substr(2, 9))
			.toUpperCase()
};
