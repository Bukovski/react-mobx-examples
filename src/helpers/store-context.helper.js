import { createContext, useContext } from 'react';


export const StoreContext = createContext({});
export const StoreProvider = StoreContext.Provider;

export const useStore = () => useContext(StoreContext);
