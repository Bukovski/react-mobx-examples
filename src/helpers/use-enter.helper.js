export const onEnterPress = (callback) => {
	return (event) => {
		if (event.key === 'Enter') {
			callback();
		}
	}
};
