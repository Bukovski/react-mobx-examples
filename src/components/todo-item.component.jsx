import React, { useState } from 'react';
import { useStore, onEnterPress } from "../helpers";


export const TodoItem = (props) => {
	const { todo } = props;
	
	const todoListStore = useStore();
	
	const [ text, setText ] = useState('');
	const [ edit, setEdit ] = useState(false);
	
	const saveText = () => {
		todo.updateText(text);
		
		setEdit(false);
		setText('');
	};
	
	return (
		<div className="todo-item">
			{
				edit
					? <li className="edit-item">
						<input
							type="text"
							onKeyDown={ onEnterPress(saveText) }
							onChange={ (event) => setText(event.target.value) }
						/>
						<i className="far fa-save save" onClick={ saveText }/>
					</li>
					: <li>
						<input
							type="checkbox"
							id={`todo_${ todo.id }`}
							defaultChecked={ todo.isDone }
							onChange={ todo.toggleIsDone }
						/>
						<label htmlFor={`todo_${ todo.id }`}>
							<span className="check" />
							{ todo.text }
						</label>
						<div className="todo-buttons">
							<i className="far fa-edit edit" onClick={ () => setEdit(true) }/>
							<i className="far fa-trash-alt delete" onClick={ () => todoListStore.removeTodo(todo) }/>
						</div>
					</li>
			}
		</div>
	)
};
