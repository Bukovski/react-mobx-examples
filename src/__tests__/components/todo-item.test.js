import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { TodoItem } from "../../components";
import { StoreProvider } from "../../helpers";



describe("TodoItem", () => {
	it("render page with props", () => {
		const todo = { id: "_sdf234lkjsdf", text: "Should be text", isDone: false  }
		
		const { container } = render(<TodoItem todo={ todo }/>);
		
		expect(screen.getByText("Should be text")).toBeInTheDocument();
		expect(screen.getByRole("checkbox")).toHaveAttribute("id", "todo__sdf234lkjsdf")
		
		expect(container.firstChild).toMatchSnapshot();
	})
	
	it('should delete item', () => {
		const todoListStore = {
			removeTodo: jest.fn()
		};
		const todo = { id: "_sdf234lkjsdf", text: "Should be text", isDone: false  }
		
		const { container } = render(<StoreProvider value={ todoListStore }>
			<TodoItem todo={ todo }/>
		</StoreProvider>);
		
		expect(todoListStore.removeTodo).toHaveBeenCalledTimes(0);

		userEvent.click(container.querySelector('i.delete'))
		expect(todoListStore.removeTodo).toHaveBeenCalledTimes(1);
	});
	
	it('should edit item', () => {
		const mockUpdateText = jest.fn();
		const todo = { id: "_sdf234lkjsdf", text: "Should be text", isDone: false, updateText: mockUpdateText  }
		
		const { container } = render(<TodoItem todo={ todo }/>);
		
		expect(mockUpdateText).toHaveBeenCalledTimes(0);
		
		userEvent.click(container.querySelector('i.edit'));
		expect(screen.getByRole("textbox")).toHaveValue("Should be text");
		
		userEvent.type(screen.getByRole("textbox"), " new text");
		expect(screen.getByRole("textbox")).toHaveValue("Should be text new text");
		
		userEvent.click(container.querySelector('i.save'));
		expect(mockUpdateText).toHaveBeenCalledTimes(1);
		
		userEvent.click(container.querySelector('i.edit'));
		userEvent.type(screen.getByRole("textbox"), " if press enter");
		expect(screen.getByRole("textbox")).toHaveValue("Should be text new text if press enter")
		
		userEvent.type(screen.getByRole("textbox"), '{enter}')
		expect(mockUpdateText).toHaveBeenCalledTimes(2);
	});
})