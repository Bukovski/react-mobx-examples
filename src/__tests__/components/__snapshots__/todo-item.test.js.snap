// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem render page with props 1`] = `
<div
  class="todo-item"
>
  <li>
    <input
      id="todo__sdf234lkjsdf"
      type="checkbox"
    />
    <label
      for="todo__sdf234lkjsdf"
    >
      <span
        class="check"
      />
      Should be text
    </label>
    <div
      class="todo-buttons"
    >
      <i
        class="far fa-edit edit"
      />
      <i
        class="far fa-trash-alt delete"
      />
    </div>
  </li>
</div>
`;
